pydis_site/apps/content/utils.py

```diff
@@ -27,8 +39,111 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+    """
+    return fetch_tags()
+
+
+def fetch_tags() -> list[Tag]:
+    """
+    Fetch tag data from the GitHub API.
+
+    The entire repository is downloaded and extracted locally because
+    getting file content would require one request per file, and can get rate-limited.
+    """
+    if settings.GITHUB_TOKEN:  # pragma: no cover
+        headers = {"Authorization": f"token {settings.GITHUB_TOKEN}"}
+    else:
+        headers = {}
+
+    tar_file = httpx.get(
+        f"{settings.GITHUB_API}/repos/python-discord/bot/tarball",
+        follow_redirects=True,
+        timeout=settings.TIMEOUT_PERIOD,
+        headers=headers,
+    )
+    tar_file.raise_for_status()
+
+    tags = []
+    with tempfile.TemporaryDirectory() as folder:
+        with tarfile.open(fileobj=BytesIO(tar_file.content)) as repo:
+            included = []
+            for file in repo.getmembers():
+                if "/bot/resources/tags" in file.path:
+                    included.append(file)
+            repo.extractall(folder, included)
+
+        for tag_file in Path(folder).rglob("*.md"):
+            tags.append(Tag(
+                name=tag_file.name.removesuffix(".md"),
+                body=tag_file.read_text(encoding="utf-8"),
+                url=f"{TAG_URL_BASE}/{tag_file.name}"
+            ))
+
+    return tags
+
+
+def get_tags() -> list[Tag]:
+    """Return a list of all tags visible to the application, from the cache or API."""
+    if settings.STATIC_BUILD:  # pragma: no cover
+        last_update = None
+    else:
+        last_update = (
+            Tag.objects.values_list("last_updated", flat=True)
+            .order_by("last_updated").first()
+        )
+
+    if last_update is None or timezone.now() >= (last_update + TAG_CACHE_TTL):
+        # Stale or empty cache
+        if settings.STATIC_BUILD:  # pragma: no cover
+            tags = get_tags_static()
+        else:
+            tags = fetch_tags()
+            Tag.objects.exclude(name__in=[tag.name for tag in tags]).delete()
+            for tag in tags:
+                tag.save()
+
+        return tags
+    else:
+        # Get tags from database
+        return Tag.objects.all()
+
+
+def get_tag(name: str) -> Tag:
+    """Return a tag by name."""
+    tags = get_tags()
+    for tag in tags:
+        if tag.name == name:
+            return tag
+
+    raise Tag.DoesNotExist()
+
+
+def get_category_pages(path: Path) -> dict[str, dict]:
     """Get all page names and their metadata at a category path."""
+    # Special handling for tags
+    if path == Path(__file__).parent / "resources/tags":
+        tags = {}
+        for tag in get_tags():
+            content = frontmatter.parse(tag.body)[1]
+            if len(content) > 100:
+                # Trim the preview to a maximum of 100 visible characters
+                # This causes some markdown to break, but we ignore that
+                content = content[:100] + "..."
```

Comment:
```suggestion
                content = content[:100] + "…"
```
😡

Reply:
Hahaha, I'm actually working on making the cropping here better, just keep your eyes peeled.

Reply:
Stale due to f2ad3eed